// quick-build-core/lib/task/main.js
var uglify2 = require('../uglify2');
var _ = require('underscore');
var semver = require('semver');//标准版本号格式检查
module.exports = function(grunt) {
	var path = require('path');
	var rDefine = /define\(\s*(['"](.+?)['"],)?/,
		rQueryHash = /[\?#].*$/,
		rModId = /([^\\\/?]+?)(\.(?:js))?([\?#].*)?$/,
		rRequire = /"(?:\\"|[^"])*"|'(?:\\'|[^'])*'|\/\*[\S\s]*?\*\/|\/(?:\\\/|[^\/\r\n])+\/(?=[^\/])|\/\/.*|\.\s*require|(?:^|[^$])\brequire\s*\(\s*(["'])(.+?)\1\s*\)/g;
	rRequireAsync = /(require\.async\(["'](.+)["'])/g;
	var Log = grunt.log;
	var xdRepsContents = {};

	grunt.registerMultiTask('build', 'build your seajs modules.', function() {
		var options = this.options({
			id: '',
			uglify:false,
			modulecfg:{}//模块信息配置,从pkg中读取
		});
		options.modulecfg = _.extend({
			uglify: {//压缩的选项
				//http://lisperator.net/uglifyjs/compress
				options: {
					global_defs: {}
				},
				//http://lisperator.net/uglifyjs/codegen 
				output: {
					indent_level: 2,
					ascii_only: true//unicode编码输出
				}
			}
		},options.modulecfg);
		var wrap_prefix = 'define(function(require, exports, module) {';
		var wrap_suffix = '});';
		//存放原始依赖与解析过后的依赖对应关系
		var idMap = {};

		var pkg = 'package.json';
		if (grunt.file.exists(pkg)) {
			pkg = grunt.file.readJSON(pkg);
			options.pkg = pkg;
		} else {
			grunt.fail.fatal('package.json not found.');
		}
		var idleading = options.id + '/' + pkg.name + '/' + pkg.version + '/';

		//将依赖（id）解析为真实（用户配置）的模块ID
		function parseRealDep(dep){
			if(options.modulecfg){
				if(_.isArray(dep)){
					return parseRealDepNames(dep);
				}else{
					return parseRealDepName(dep);
				}
			}else{
				return dep;
			}
		}
		//将一个依赖（id）数组解析为真实（用户配置）的模块ID数组
		function parseRealDepNames(depNames){
			return _.map(depNames,function(depName){
				return parseRealDepName(depName);
			});
		}
		//将一个依赖（id）解析为真实（用户配置）的模块ID
		function parseRealDepName(depName){
			var cfg = options.modulecfg[depName];
			if(_.isEmpty(cfg)){
				return depName;
			}else if(_.isString(cfg)){
				if (!semver.valid(cfg)) {//不是标准版本号
					return cfg;
				}else{
					//只给了一个版本号，则模块前缀默认与当前模块前缀一致
					return options.id+'/'+depName+'/'+cfg+'/index';
				}
			}else if(_.isArray(cfg) && cfg.length==2){
				return cfg[0]+'/'+depName+'/'+cfg[1]+'/index';
			}
		}

		function parse(id, $2, callback) {
			var result = id,depId, depOrigId, depPathResult,firstStr;
			if ($2 && $2.slice(0, 4) !== 'http') {
				// { id: 'a', path: './src/a', extName: ''}
				depPathResult = modPathResolve(options, $2);
				depOrigId = depPathResult.path;
				depId = idMap[depOrigId];
				if (String(depOrigId).slice(0, 1) == '.') {
					result = callback(parseDepsText(depId));
				} else {
					result = callback(depId);
				}
			}
			return result;
		}

		this.files.forEach(function(file) {
			var contents = file.src.filter(function(filepath) {
				if (!grunt.file.exists(filepath)) {
					Log.warn('Source file "' + filepath + '" not found.');
					return false;
				} else {
					return true;
				}
			}).map(function(filepath) {
				var content = grunt.file.read(filepath);
				var fileExt = path.extname(filepath);
				var fileName = path.basename(filepath, fileExt);
				var dirName = path.dirname(filepath);
				var depsMap = {};

				switch (fileExt) {
					case '.tpl':
						content = "return '" + content.replace(/\s/g, ' ') + "'";
						break;
				}

				if (!rDefine.test(content)) {
					content = [wrap_prefix, content, wrap_suffix].join('\n');
				}

				//解析出同步加载的依赖数组
				var deps = pullDeps(options, rRequire, content);
				//解析出异步加载的依赖数组，并于同步的合并在一起
				deps = deps.concat(pullDeps(options, rRequireAsync, content));

				//解析依赖关系数组
				var depNamesAsync = deps.map(function(dep) {
					var beginStr = dep.origId.slice(0, 1);
					var besePath = idleading + dirName;
					//以相对路径开头需要将basePath与id组合起来
					idMap[dep.origId] = depsMap[dep.id]
					= path.join((beginStr == '.' ? besePath : ''), dep.origId);
					return depsMap[dep.id];
				});

				//define
				content = content.replace(rDefine, function() {
					var id = idleading + (dirName == '.' ?
						'' : (path.normalize(dirName) == path.normalize(file.orig.cwd)) ?
						'' : formatDepsText(path.relative(file.orig.cwd, dirName)) + '/') + fileName;
					//(path.relative(file.orig.cwd,dirName)).replace(/\\/g,'/')
					//计算出相对于基本路径的路径，可能存在\符号，直接替换成/
					//cwd的路径与输出文件的基本路径一致，则返回'',否则返回基础文件夹+文件名
					depNames = parseRealDep(depNamesAsync);
					var tmp = [];
					depNames.forEach(function(id){
						var lastChar = id.slice(-1);
						if(lastChar!='/'){
							tmp.push(id);
						}
					});
					depNames = tmp;
					return "define('" + id + "',"+ (depNames.length ? ("['"+parseDepsText(depNames.join("','"))+"'],"):(""));
				});

				//同步加载的
				content = content.replace(rRequire, function(id, _, $2) {
					return parse(id, $2, function(depId) {
						depId = parseRealDep(depId);
						return "require('" + depId + "')"
					});
				});

				//异步加载的
				// content = content.replace(rRequireAsync, function(id, _, $2) {
				// 	return parse(id, $2, function(depId) {
				// 		depId = parseRealDep(depId);
				// 		return "require.async('" + depId + "'"
				// 	});
				// });

				//返回处理结果
				return content;
			}).join('\n');

			//文件的路径相对与cwd的路径
			var _dest = path.join(file.orig.dest,idleading,path.relative(file.orig.cwd,file.src[0]));
			try {
				if (options.uglify) {//混淆
					contents = uglify2.min(contents,options.modulecfg.uglify).code;
				}
			} catch (e) {
				Log.error('ERROR：（配置可能有误）压缩代码失败：' + e);
			}

			var fileExt = path.extname(_dest);
			switch (fileExt) {
				case '.tpl':
					_dest = _dest.replace(fileExt, '.js');
					break;
			}
			grunt.file.write(_dest, contents);
			Log.writeln('File "' + _dest + '" created.');
		});
	});

	//格式化
	function formatDepsText(depsText) {
		//转换\为/，解决win下的路径BUG
		return depsText.replace(/\\/g, '/');
	}

	//将.tpl后缀去掉，tpl模版在转换的过程中已经变成了js模块了
	function parseDepsText(depsText) {
		return formatDepsText(depsText).replace(/\.tpl/g, '');
	}

	function pullDeps(options, reg, contents) {
		var deps = [],
			matches, origId;
		reg.lastIndex = 0;
		while ((matches = reg.exec(contents)) !== null) {
			origId = matches[2];
			if (origId && origId.slice(0, 4) !== 'http') {
				depPathResult = modPathResolve(options, origId);
				console.log(depPathResult.id);
				deps.push({
					id: depPathResult.id,
					origId: depPathResult.path,
					extName: depPathResult.extName
				});
			}
		}
		return deps;
	}

	/*
	 * 解析模块标识
	 * param { Object } 配置参数
	 * param { String } 模块标识
	 * return { Object } filePath: 过滤query和hash后的模块标识,id: 模块id,extName: 模块后缀
	 */
	function modPathResolve(options, filePath) {
		// 过滤query(?)和hash(#)
		filePath = filePath.replace(rQueryHash, '');
		var t = filePath.match(rModId);
		var id = t && t[1],
			extName = path.extname(filePath);
		if (extName && extName === '.js') {
			id = id.replace(extName, '');
		}
		return {
			id: id,
			path: filePath,
			extName: extName
		};
	}

	return grunt;
}
